server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java

```diff
@@ -273,6 +301,40 @@ public static MetadataVersion latest() {
         return VALUES[VALUES.length - 1];
     }
 
+    public Optional<MetadataVersion> previous() {
+        int idx = this.ordinal();
+        if (idx > 2) {
+            return Optional.of(MetadataVersion.values()[idx - 1]);
```

Comment:
```suggestion
            return Optional.of(VALUES[idx]);
```
The indexing is a bit unintuitive, I wonder if you might just want to store all of `values()` (including `UNINITIALIZED`) in `VALUES` just to make this more clear?

Reply:
I'm going to reorganize this code a bit to make it more clear.